django/contrib/postgres/signals.py

```diff
@@ -32,30 +30,51 @@ def get_citext_oids(connection_alias):
     return get_type_oids(connection_alias, "citext")
 
 
-def register_type_handlers(connection, **kwargs):
-    if connection.vendor != "postgresql" or connection.alias == NO_DB_ALIAS:
-        return
-
-    oids, array_oids = get_hstore_oids(connection.alias)
-    # Don't register handlers when hstore is not available on the database.
-    #
-    # If someone tries to create an hstore field it will error there. This is
-    # necessary as someone may be using PSQL without extensions installed but
-    # be using other features of contrib.postgres.
-    #
-    # This is also needed in order to create the connection in order to install
-    # the hstore extension.
-    if oids:
-        register_hstore(
-            connection.connection, globally=True, oid=oids, array_oid=array_oids
-        )
+if is_psycopg3:
+    from psycopg.types import TypeInfo, hstore
 
-    oids, citext_oids = get_citext_oids(connection.alias)
-    # Don't register handlers when citext is not available on the database.
-    #
-    # The same comments in the above call to register_hstore() also apply here.
-    if oids:
-        array_type = psycopg2.extensions.new_array_type(
-            citext_oids, "citext[]", psycopg2.STRING
-        )
-        psycopg2.extensions.register_type(array_type, None)
+    def register_type_handlers(connection, **kwargs):
+        if connection.vendor != "postgresql" or connection.alias == NO_DB_ALIAS:
+            return
+
+        oids, array_oids = get_hstore_oids(connection.alias)
+        for oid, array_oid in zip(oids, array_oids):
+            ti = TypeInfo("hstore", oid, array_oid)
+            hstore.register_hstore(ti, connection.connection)
+
+        _, citext_oids = get_citext_oids(connection.alias)
+        for array_oid in citext_oids:
+            ti = TypeInfo("citext", 0, array_oid)
```

Comment:
Is there any reason why we don't register the `citext` oid? 
```
florian=# SELECT oid, typarray FROM pg_type WHERE typname = 'citext';
   oid   | typarray 
---------+----------
 8611888 |  8611893
```

Reply:
:shrug:

Reply:
@dvarrazzo Any idea why we are using `0` here to register citext? I realize the old code had it as well (or rather just registered `citext[]`) so I am not sure what would be the better/correct way.

Reply:
Ok `citext` seems to be somewhat special since it mostly behaves like a string :)

```
diff --git a/django/contrib/postgres/signals.py b/django/contrib/postgres/signals.py
index a3816d3d30..06d7eec9c5 100644
--- a/django/contrib/postgres/signals.py
+++ b/django/contrib/postgres/signals.py
@@ -42,9 +42,9 @@ if is_psycopg3:
             ti = TypeInfo("hstore", oid, array_oid)
             hstore.register_hstore(ti, connection.connection)
 
-        _, citext_oids = get_citext_oids(connection.alias)
-        for array_oid in citext_oids:
-            ti = TypeInfo("citext", 0, array_oid)
+        oids, citext_oids = get_citext_oids(connection.alias)
+        for oid, array_oid in zip(oids, citext_oids):
+            ti = TypeInfo("citext", oid, array_oid)
             ti.register(connection.connection)
 
 else:
```

This change seems to pass the testsuite as well, so for consistency I wonder if we wouldn't want to register the normal type as well.

Reply:
The reason for it to be 0 was probably to allow to insert the value or use it in contexts where citext is not expected, the same reason for which text uses the oid 0.

Using oid 0 gives psycopg 3 queries the same behaviour of psycopg2. If the oid is specified, bets are off for customer queries not covered by the test suite.

I thought there was already some test failing to "warn" about it, but probably I'm wrong. My suggestion is to use 0 as oid, but not as strongly as for text.

Reply:
Also all `ci`-fields are deprecated anyway so I would keep the current behavior.